crates/sui-indexer/src/store/pg_partition_manager.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use diesel::sql_types::VarChar;
+use diesel::{QueryableByName, RunQueryDsl};
+use std::collections::BTreeMap;
+use std::str::FromStr;
+use tracing::info;
+
+use crate::handlers::EpochToCommit;
+use crate::store::diesel_marco::{read_only_blocking, transactional_blocking};
+use crate::IndexerError;
+use crate::PgConnectionPool;
+
+const GET_PARTITION_SQL: &str = r#"
+SELECT parent.relname                           AS table_name,
+       MAX(SUBSTRING(child.relname FROM '\d$')) AS last_partition
+FROM pg_inherits
+         JOIN pg_class parent ON pg_inherits.inhparent = parent.oid
+         JOIN pg_class child ON pg_inherits.inhrelid = child.oid
+         JOIN pg_namespace nmsp_parent ON nmsp_parent.oid = parent.relnamespace
+         JOIN pg_namespace nmsp_child ON nmsp_child.oid = child.relnamespace
+WHERE parent.relkind = 'p'
+GROUP BY table_name;
+"#;
+
+#[derive(Clone)]
+pub struct PgPartitionManager {
+    cp: PgConnectionPool,
+}
+
+#[derive(Clone)]
+pub struct EpochPartitionData {
+    last_epoch: u64,
+    next_epoch: u64,
+    last_epoch_start_cp: u64,
+    next_epoch_start_cp: u64,
+}
+
+impl From<EpochToCommit> for EpochPartitionData {
+    fn from(epoch: EpochToCommit) -> Self {
+        let last_epoch = epoch.last_epoch.as_ref().map(|e| e.epoch).unwrap_or(0);
+        let next_epoch = epoch.new_epoch.epoch;
+        let last_epoch_start_cp = epoch.last_epoch.map(|e| e.first_checkpoint_id).unwrap_or(0);
```

Comment:
unfortunately this won't work because at the end of epoch X, we won't know its starting checkpoint id unless read from db: https://github.com/MystenLabs/sui/blob/main/crates/sui-indexer/src/types_v2.rs#L158